{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"Pathway Analysis\"\n",
    "author: \"Javier Millan Acosta\"\n",
    "date: 2024-11-28\n",
    "---\n",
    "\n",
    "## 1. Introduction\n",
    "\n",
    " > Note: adapted for Python from [Pathway-Analysis.Rmd](https://github.com/wikipathways/rWikiPathways/blob/devel/vignettes/Pathway-Analysis.Rmd).\n",
    "\n",
    "This notebook will cover a wide range of analytical and visualization techniques involved in a typical pathway analysis. The **Overview** section will go into more detail on the particulars, but note that this vignette is designed to be modular and carefully considered. Please do not simply run the entire script and expect to get anything meaningful from the final output. This is an instructional device, ideal for guided workshops.\n",
    "\n",
    "### 1.1. Installation\n",
    "\n",
    "#### Packages\n",
    "The following packages can be installed with the `pip` package installer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install -r requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import py4cytoscape as p4c\n",
    "import pandas as pd\n",
    "from biomart import BiomartServer\n",
    "import pywikipathways as pwpw\n",
    "import numpy as np\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import json\n",
    "import requests\n",
    "from io import StringIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2. Cytoscape installation and setup\n",
    "\n",
    "The `py4cytoscape` package is used to connect with **Cytoscape**. So you will also need to install and launch Cytoscape.\n",
    "\n",
    "* Download the latest Cytoscape from https://cytoscape.org/download.html \n",
    "* Complete installation wizard\n",
    "* Launch Cytoscape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You are connected to Cytoscape!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'You are connected to Cytoscape!'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p4c.cytoscape_ping()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this vignette, you'll also need a couple *apps* for Cytoscape. With Cytoscape running, you can install each of these from the Cytoscape App Store with a single click:\n",
    "\n",
    "* http://apps.cytoscape.org/apps/wikipathways\n",
    "* http://apps.cytoscape.org/apps/cytargetlinker \n",
    "\n",
    "If you are running Cytoscape 3.7.0 or above, you can simply run these commands:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p4c.install_app(\"CyTargetLinker\")\n",
    "p4c.install_app(\"WikiPathways\")\n",
    "p4c.install_app(\"stringApp\")\n",
    "p4c.install_app(\"enrichmentMap\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Overview\n",
    "\n",
    "In this vignette, we will be performing functional enrichment analysis on a differential gene expression dataset. The dataset compares the expression of transcripts in lung cancer biopses versus normal tissue. Differential expression analysis has already been performed, generating log2foldchange and P-values for each gene. The enrichment analysis will be performed against Gene Ontology, as an introduction to the most common type of enrichment, commonly referred to as *GO Analysis*. This will serve as the foundation for more advanced enrichment analysis against a pathway database, which is called *Pathway Analysis*.\n",
    "\n",
    "Working with pathways opens up unique analysis and visualization options. We will query WikiPathways for relevant content and import pathway models into Cytoscape. In Cytoscape, we will perform data overlays, add drug interactions and generate high-quality images for publication.\n",
    "\n",
    "### 2.2. Dataset\n",
    "\n",
    "The format of this lung cancer dataset should look familiar to anyone who has worked with differential gene expression results. It contains columns of gene identifiers (Ensembl IDs), gene symbols, log2FC values, P-Values and adjusted P-Values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8823</th>\n",
       "      <td>ENSG00000146521</td>\n",
       "      <td>LINC01558</td>\n",
       "      <td>0.036464</td>\n",
       "      <td>0.962679</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26445</th>\n",
       "      <td>ENSG00000233477</td>\n",
       "      <td>PPIAP38</td>\n",
       "      <td>-0.661354</td>\n",
       "      <td>0.835649</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33145</th>\n",
       "      <td>ENSG00000258984</td>\n",
       "      <td>UBE2F-SCLY</td>\n",
       "      <td>1.362593</td>\n",
       "      <td>0.190051</td>\n",
       "      <td>0.463952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneID    GeneName    log2FC   P.Value  adj.P.Value\n",
       "8823   ENSG00000146521   LINC01558  0.036464  0.962679     1.000000\n",
       "26445  ENSG00000233477     PPIAP38 -0.661354  0.835649     1.000000\n",
       "33145  ENSG00000258984  UBE2F-SCLY  1.362593  0.190051     0.463952"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_expr = pd.read_excel(os.path.join(\"extdata\", \"lung-cancer-data.xlsx\"))\n",
    "lung_expr = lung_expr[\n",
    "    lung_expr[\"GeneName\"].apply(lambda x: isinstance(x, str))\n",
    "]  # Need to clean datetime types added by Excel\n",
    "lung_expr.sample(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3. Up- and downregulated genes\n",
    "\n",
    "Now let's prepare up- and down-regulated gene lists using some conventional criteria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "P_CUTOFF = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "lung_expr[\"de\"] = np.where(\n",
    "    (lung_expr[\"log2FC\"] > 1) & (lung_expr[\"adj.P.Value\"] < P_CUTOFF),\n",
    "    \"upregulated\",\n",
    "    np.where(\n",
    "    (lung_expr[\"log2FC\"] < -1) & (lung_expr[\"adj.P.Value\"] < P_CUTOFF), \n",
    "    \"downregulated\", \n",
    "    'bkgd'),)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Value</th>\n",
       "      <th>de</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>36814</td>\n",
       "      <td>36814</td>\n",
       "      <td>36814.000000</td>\n",
       "      <td>3.681400e+04</td>\n",
       "      <td>36814.000000</td>\n",
       "      <td>36814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>36786</td>\n",
       "      <td>36766</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>ENSG00000277739</td>\n",
       "      <td>RNA5-8S4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bkgd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.738960</td>\n",
       "      <td>3.990082e-01</td>\n",
       "      <td>0.584830</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.777559</td>\n",
       "      <td>3.646529e-01</td>\n",
       "      <td>0.339227</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-8.742461</td>\n",
       "      <td>3.580000e-08</td>\n",
       "      <td>0.000853</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.139523</td>\n",
       "      <td>6.126635e-02</td>\n",
       "      <td>0.262821</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.268689</td>\n",
       "      <td>2.912378e-01</td>\n",
       "      <td>0.560967</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.479487</td>\n",
       "      <td>7.396475e-01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.717148</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GeneID  GeneName        log2FC       P.Value   adj.P.Value  \\\n",
       "count             36814     36814  36814.000000  3.681400e+04  36814.000000   \n",
       "unique            36786     36766           NaN           NaN           NaN   \n",
       "top     ENSG00000277739  RNA5-8S4           NaN           NaN           NaN   \n",
       "freq                  6         4           NaN           NaN           NaN   \n",
       "mean                NaN       NaN      0.738960  3.990082e-01      0.584830   \n",
       "std                 NaN       NaN      1.777559  3.646529e-01      0.339227   \n",
       "min                 NaN       NaN     -8.742461  3.580000e-08      0.000853   \n",
       "25%                 NaN       NaN     -0.139523  6.126635e-02      0.262821   \n",
       "50%                 NaN       NaN      0.268689  2.912378e-01      0.560967   \n",
       "75%                 NaN       NaN      1.479487  7.396475e-01      1.000000   \n",
       "max                 NaN       NaN     13.717148  1.000000e+00      1.000000   \n",
       "\n",
       "           de  \n",
       "count   36814  \n",
       "unique      3  \n",
       "top      bkgd  \n",
       "freq    35791  \n",
       "mean      NaN  \n",
       "std       NaN  \n",
       "min       NaN  \n",
       "25%       NaN  \n",
       "50%       NaN  \n",
       "75%       NaN  \n",
       "max       NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_expr.describe(include=\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: >"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZIAAAD4CAYAAADGmmByAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8hTgPZAAAACXBIWXMAAAsTAAALEwEAmpwYAAAZUklEQVR4nO3df5RfdX3n8efLABqxCgrOYQEbzjHtbpQaZQRctWeqLQTaHrC1Li4r0XKMrdi1e+hW7J4tVsoebQ/1FKq0UVOCS0WKZZPa2JhFpog9QIIiISBLinFJDsipCejoihv3vX98PyPfxpnMzPfOTGbw+Tjne+Z+3/fzufdzc2++r7k/ZiZVhSRJg3rGoR6AJGlxM0gkSZ0YJJKkTgwSSVInBokkqZPDDvUABnXMMcfUsmXLBur7ne98hyOPPHJ2ByQ1Hl+aS12Pr7vuuuufq+rYWRzS4g2SZcuWsW3btoH6jo6OMjIyMrsDkhqPL82lrsdXkq/P3mh6vLQlSerEIJEkdWKQSJI6MUgkSZ0YJJKkTgwSSVInBokkqRODRJLUiUEiSepk0f5kexfb9zzBWy/5u3lf764P/OK8r1OS5tqUZyRJnpXkziRfSbIjyR+0+jVJvpbk7vZa2epJcmWSnUnuSfKKvmWtTvJge63uq5+SZHvrc2WSzMG2SpLmwHTOSJ4EXldVY0kOB25L8tk27z9X1Y0HtD8LWN5epwFXA6cleT5wKTAMFHBXko1Vta+1eTtwB7AJWAV8FknSgjflGUn1jLW3h7fXwf7Q+znAta3f7cBRSY4DzgS2VNXeFh5bgFVt3nOr6vbq/QH5a4FzB98kSdJ8mtY9kiRLgLuAFwMfrqo7kvwmcHmS3wduBi6pqieB44GH+7rvbrWD1XdPUJ9oHGuANQBDQ0OMjo5OZ/g/YmgpXHzy/oH6djHoeLW4jI2Nua81Zxbi8TWtIKmqHwArkxwF3JTkpcB7gUeBI4C1wHuA98/ROMfHsbati+Hh4Rr0Vylfdd0Grtg+/88Z7Dp/ZN7Xqfnnr5HXXFqIx9eMHv+tqseBW4BVVfVIu3z1JPCXwKmt2R7gxL5uJ7TaweonTFCXJC0C03lq69h2JkKSpcAvAF9t9zZoT1idC9zbumwELmhPb50OPFFVjwCbgTOSHJ3kaOAMYHOb960kp7dlXQBsmM2NlCTNnelc3zkOWN/ukzwDuKGqPpPk80mOBQLcDfxGa78JOBvYCXwXeBtAVe1NchmwtbV7f1XtbdPvBK4BltJ7WssntiRpkZgySKrqHuDlE9RfN0n7Ai6aZN46YN0E9W3AS6caiyRp4fFXpEiSOjFIJEmdGCSSpE4MEklSJwaJJKkTg0SS1IlBIknqxCCRJHVikEiSOjFIJEmdGCSSpE4MEklSJwaJJKkTg0SS1IlBIknqxCCRJHVikEiSOjFIJEmdGCSSpE4MEklSJ1MGSZJnJbkzyVeS7EjyB61+UpI7kuxM8qkkR7T6M9v7nW3+sr5lvbfVH0hyZl99VavtTHLJHGynJGmOTOeM5EngdVX1MmAlsCrJ6cAHgQ9V1YuBfcCFrf2FwL5W/1BrR5IVwHnAS4BVwEeSLEmyBPgwcBawAnhzaytJWgSmDJLqGWtvD2+vAl4H3Njq64Fz2/Q57T1t/uuTpNWvr6onq+prwE7g1PbaWVUPVdX3getbW0nSInDYdBq1s4a7gBfTO3v4J+DxqtrfmuwGjm/TxwMPA1TV/iRPAC9o9dv7Ftvf5+ED6qdNMo41wBqAoaEhRkdHpzP8HzG0FC4+ef/UDWfZoOPV4jI2Nua+1pxZiMfXtIKkqn4ArExyFHAT8K/nclAHGcdaYC3A8PBwjYyMDLScq67bwBXbp7Xps2rX+SPzvk7Nv9HRUQY9NqWpLMTja0ZPbVXV48AtwKuAo5KMfxqfAOxp03uAEwHa/OcB3+yvH9BnsrokaRGYzlNbx7YzEZIsBX4BuJ9eoLyxNVsNbGjTG9t72vzPV1W1+nntqa6TgOXAncBWYHl7CuwIejfkN87CtkmS5sF0ru8cB6xv90meAdxQVZ9Jch9wfZI/BL4MfLy1/zjwiSQ7gb30goGq2pHkBuA+YD9wUbtkRpJ3AZuBJcC6qtoxa1soSZpTUwZJVd0DvHyC+kP0nrg6sP494NcmWdblwOUT1DcBm6YxXknSAuNPtkuSOjFIJEmdGCSSpE4MEklSJwaJJKkTg0SS1IlBIknqxCCRJHVikEiSOjFIJEmdGCSSpE4MEklSJwaJJKkTg0SS1IlBIknqxCCRJHVikEiSOjFIJEmdGCSSpE4MEklSJ1MGSZITk9yS5L4kO5K8u9Xfl2RPkrvb6+y+Pu9NsjPJA0nO7KuvarWdSS7pq5+U5I5W/1SSI2Z7QyVJc2M6ZyT7gYuragVwOnBRkhVt3oeqamV7bQJo884DXgKsAj6SZEmSJcCHgbOAFcCb+5bzwbasFwP7gAtnafskSXNsyiCpqkeq6ktt+tvA/cDxB+lyDnB9VT1ZVV8DdgKnttfOqnqoqr4PXA+ckyTA64AbW//1wLkDbo8kaZ4dNpPGSZYBLwfuAF4NvCvJBcA2emct++iFzO193XbzVPA8fED9NOAFwONVtX+C9geufw2wBmBoaIjR0dGZDP+HhpbCxSfvn7rhLBt0vFpcxsbG3NeaMwvx+Jp2kCR5DvBp4Ler6ltJrgYuA6p9vQL49TkZZVNVa4G1AMPDwzUyMjLQcq66bgNXbJ9Rhs6KXeePzPs6Nf9GR0cZ9NiUprIQj69pfZomOZxeiFxXVX8DUFXf6Jv/UeAz7e0e4MS+7ie0GpPUvwkcleSwdlbS316StMBN56mtAB8H7q+qP+mrH9fX7A3AvW16I3BekmcmOQlYDtwJbAWWtye0jqB3Q35jVRVwC/DG1n81sKHbZkmS5st0zkheDbwF2J7k7lb7PXpPXa2kd2lrF/AOgKrakeQG4D56T3xdVFU/AEjyLmAzsARYV1U72vLeA1yf5A+BL9MLLknSIjBlkFTVbUAmmLXpIH0uBy6foL5pon5V9RC9p7okSYuMP9kuSerEIJEkdWKQSJI6MUgkSZ0YJJKkTgwSSVInBokkqRODRJLUiUEiSerEIJEkdWKQSJI6MUgkSZ0YJJKkTgwSSVInBokkqRODRJLUiUEiSerEIJEkdWKQSJI6MUgkSZ1MGSRJTkxyS5L7kuxI8u5Wf36SLUkebF+PbvUkuTLJziT3JHlF37JWt/YPJlndVz8lyfbW58okmYuNlSTNvumckewHLq6qFcDpwEVJVgCXADdX1XLg5vYe4CxgeXutAa6GXvAAlwKnAacCl46HT2vz9r5+q7pvmiRpPkwZJFX1SFV9qU1/G7gfOB44B1jfmq0Hzm3T5wDXVs/twFFJjgPOBLZU1d6q2gdsAVa1ec+tqturqoBr+5YlSVrgDptJ4yTLgJcDdwBDVfVIm/UoMNSmjwce7uu2u9UOVt89QX2i9a+hd5bD0NAQo6OjMxn+Dw0thYtP3j9Q3y4GHa8Wl7GxMfe15sxCPL6mHSRJngN8GvjtqvpW/22MqqokNQfj+xeqai2wFmB4eLhGRkYGWs5V123giu0zytBZsev8kXlfp+bf6Ogogx6b0lQW4vE1rae2khxOL0Suq6q/aeVvtMtStK+Ptfoe4MS+7ie02sHqJ0xQlyQtAtN5aivAx4H7q+pP+mZtBMafvFoNbOirX9Ce3jodeKJdAtsMnJHk6HaT/Qxgc5v3rSSnt3Vd0LcsSdICN53rO68G3gJsT3J3q/0e8AHghiQXAl8H3tTmbQLOBnYC3wXeBlBVe5NcBmxt7d5fVXvb9DuBa4ClwGfbS5K0CEwZJFV1GzDZz3W8foL2BVw0ybLWAesmqG8DXjrVWCRJC48/2S5J6sQgkSR1YpBIkjoxSCRJnRgkkqRODBJJUicGiSSpE4NEktSJQSJJ6sQgkSR1YpBIkjoxSCRJnRgkkqRODBJJUicGiSSpE4NEktSJQSJJ6sQgkSR1YpBIkjoxSCRJnUwZJEnWJXksyb19tfcl2ZPk7vY6u2/ee5PsTPJAkjP76qtabWeSS/rqJyW5o9U/leSI2dxASdLcms4ZyTXAqgnqH6qqle21CSDJCuA84CWtz0eSLEmyBPgwcBawAnhzawvwwbasFwP7gAu7bJAkaX5NGSRVdSuwd5rLOwe4vqqerKqvATuBU9trZ1U9VFXfB64HzkkS4HXAja3/euDcmW2CJOlQOqxD33cluQDYBlxcVfuA44Hb+9rsbjWAhw+onwa8AHi8qvZP0P5HJFkDrAEYGhpidHR0oIEPLYWLT94/dcNZNuh4tbiMjY25rzVnFuLxNWiQXA1cBlT7egXw67M1qMlU1VpgLcDw8HCNjIwMtJyrrtvAFdu7ZOhgdp0/Mu/r1PwbHR1l0GNTmspCPL4G+jStqm+MTyf5KPCZ9nYPcGJf0xNajUnq3wSOSnJYOyvpby9JWgQGevw3yXF9b98AjD/RtRE4L8kzk5wELAfuBLYCy9sTWkfQuyG/saoKuAV4Y+u/GtgwyJgkSYfGlGckST4JjADHJNkNXAqMJFlJ79LWLuAdAFW1I8kNwH3AfuCiqvpBW867gM3AEmBdVe1oq3gPcH2SPwS+DHx8tjZOkjT3pgySqnrzBOVJP+yr6nLg8gnqm4BNE9QfovdUlyRpEfIn2yVJnRgkkqRODBJJUicGiSSpE4NEktSJQSJJ6sQgkSR1YpBIkjoxSCRJnRgkkqRODBJJUicGiSSpE4NEktSJQSJJ6sQgkSR1YpBIkjoxSCRJnRgkkqRODBJJUicGiSSpkymDJMm6JI8lubev9vwkW5I82L4e3epJcmWSnUnuSfKKvj6rW/sHk6zuq5+SZHvrc2WSzPZGSpLmznTOSK4BVh1QuwS4uaqWAze39wBnAcvbaw1wNfSCB7gUOA04Fbh0PHxam7f39TtwXZKkBWzKIKmqW4G9B5TPAda36fXAuX31a6vnduCoJMcBZwJbqmpvVe0DtgCr2rznVtXtVVXAtX3LkiQtAocN2G+oqh5p048CQ236eODhvna7W+1g9d0T1CeUZA29Mx2GhoYYHR0dbPBL4eKT9w/Ut4tBx6vFZWxszH2tObMQj69Bg+SHqqqS1GwMZhrrWgusBRgeHq6RkZGBlnPVdRu4YnvnTZ+xXeePzPs6Nf9GR0cZ9NiUprIQj69Bn9r6RrssRfv6WKvvAU7sa3dCqx2sfsIEdUnSIjFokGwExp+8Wg1s6Ktf0J7eOh14ol0C2wyckeTodpP9DGBzm/etJKe3p7Uu6FuWJGkRmPL6TpJPAiPAMUl203v66gPADUkuBL4OvKk13wScDewEvgu8DaCq9ia5DNja2r2/qsZv4L+T3pNhS4HPtpckaZGYMkiq6s2TzHr9BG0LuGiS5awD1k1Q3wa8dKpxSJIWJn+yXZLUiUEiSerEIJEkdWKQSJI6MUgkSZ0YJJKkTgwSSVInBokkqRODRJLUiUEiSerEIJEkdWKQSJI6MUgkSZ0YJJKkTgwSSVInBokkqRODRJLUiUEiSerEIJEkdWKQSJI66RQkSXYl2Z7k7iTbWu35SbYkebB9PbrVk+TKJDuT3JPkFX3LWd3aP5hkdbdNkiTNp9k4I/m5qlpZVcPt/SXAzVW1HLi5vQc4C1jeXmuAq6EXPMClwGnAqcCl4+EjSVr45uLS1jnA+ja9Hji3r35t9dwOHJXkOOBMYEtV7a2qfcAWYNUcjEuSNAcO69i/gM8lKeAvqmotMFRVj7T5jwJDbfp44OG+vrtbbbL6j0iyht7ZDENDQ4yOjg406KGlcPHJ+wfq28Wg49XiMjY25r7WnFmIx1fXIHlNVe1J8kJgS5Kv9s+sqmohMytaUK0FGB4erpGRkYGWc9V1G7hie9dNn7ld54/M+zo1/0ZHRxn02JSmshCPr06XtqpqT/v6GHATvXsc32iXrGhfH2vN9wAn9nU/odUmq0uSFoGBgyTJkUl+YnwaOAO4F9gIjD95tRrY0KY3Ahe0p7dOB55ol8A2A2ckObrdZD+j1SRJi0CX6ztDwE1JxpfzV1X190m2AjckuRD4OvCm1n4TcDawE/gu8DaAqtqb5DJga2v3/qra22FckqR5NHCQVNVDwMsmqH8TeP0E9QIummRZ64B1g45FknTo+JPtkqRODBJJUicGiSSpE4NEktSJQSJJ6sQgkSR1YpBIkjoxSCRJnRgkkqRODBJJUicGiSSpE4NEktSJQSJJ6sQgkSR1YpBIkjoxSCRJnRgkkqRODBJJUicGiSSpk4H/ZrukiW3f8wRvveTvDsm6d33gFw/JevXjbcGckSRZleSBJDuTXHKoxyNJmp4FcUaSZAnwYeAXgN3A1iQbq+q+QzsySZrYskN01nnNqiMPyXoPZqGckZwK7Kyqh6rq+8D1wDmHeEySpGlYEGckwPHAw33vdwOnHdgoyRpgTXs7luSBAdd3DPDPA/YdWD4432vUIXJIji/wGPtx8HMf7Hx8/eRsjWXcQgmSaamqtcDarstJsq2qhmdhSNKP8PjSXFqIx9dCubS1Bzix7/0JrSZJWuAWSpBsBZYnOSnJEcB5wMZDPCZJ0jQsiEtbVbU/ybuAzcASYF1V7ZjDVXa+PCYdhMeX5tKCO75SVYd6DJKkRWyhXNqSJC1SBokkqZOnRZAkWZbk3gnqu5IcM5vL1MKW5H1JfudQj2Mmkrw1yZ9N0WYkyb8dYNkD/x/Q4pLkmiRvnKLNW5P8qxkud8rPwqdFkEjzKcmheEhlBJhxkGjhOkTH0VuBGQXJdDydguSwJNcluT/JjUmePT4jydIkn03y9vb+v7ZfEHlbkk+Ofweb5JQkX0nyFeCiQ7QdmqEk/yXJ/0pyG/DTrbYyye1J7klyU5Kjk7wwyV1t/suSVJIXtff/lOTZ7bu6K5P8Y5KHxr/Da2cEX0iyEbgvyZIkf5xka1vHO1q7ZyT5SJKvJtmSZFPfMn54dpBkOMnoBNvyy0nuSPLlJP8zyVCSZcBvAP8pyd1JXpvk2CSfbuvfmuTVrf8LknwuyY4kHwMyp//4P+YO/G49ye+0s+LRJH/a9te9SU5t89+X5BNJvgh84iD78dh2/OxI8rEkX09yzGTrm2Bcv9+Wd2+Stel5IzAMXNfGtbR95v1DkruSbE5yXOs/o8/Cp1OQ/DTwkar6N8C3gHe2+nOAvwU+WVUfTfJK4FeBlwFn0fuHHfeXwG9V1cvmb9jqIskp9H7uaCVwNvDKNuta4D1V9TPAduDSqnoMeFaS5wKvBbYBr03yk8BjVfXd1vc44DXALwEf6FvdK4B3V9VPARcCT1TVK9s6357kJOBXgGXACuAtwKtmuEm3AadX1cvp/c65362qXcCfAx+qqpVV9QXgT9v78eP5Y63/pcBtVfUS4CbgRTNcv2bPs6tqJb3PonV99RXAz1fVmzn4fvx82483MvP9+GdV9cqqeimwFPilqrqR3jF/fhvXfuAq4I1VdUob4+Wt/4w+CxfEz5HMkoer6ott+r8D/7FNbwD+qKqua+9fDWyoqu8B30vytwBJjgKOqqpbW7tP0AsaLWyvBW4aD4F2xnAkvX35D63NeuCv2/Q/0jsGfhb4b8Aqet+1f6Fvmf+jqv4fvTOPob76nVX1tTZ9BvAzeeqa9POA5fQC6K9b/0eT3DLD7TkB+FT7zvAI4GuTtPt5YEXywxOO5yZ5TtuuXwGoqr9Lsm+G69fs+SRAVd2a5LntMwZgY1X9nzY92X58DfCG1v/vB9iPP5fkd4FnA88HdtD7hrrfTwMvBba09S8BHhnks/DpFCQH/kDM+PsvAquS/FX5QzOCW+mFz0/S+ybjPfSOlf7fCf5k33T/paHvHFD/rara3L/wJGcfZN37eeoqwLMmaXMV8CdVtTHJCPC+Sdo9g96Zy/cOWP9BVq850L9P4V/u18k+k/qPo5nux4Otb7zvs4CPAMNV9XC79DXR8RZgR1W96oD+R0228sk8nS5tvSjJ+D/Iv6d3iQDg94F99P7eCfSC5ZeTPKsl/y8BVNXjwONJXtPanT8vo1ZXtwLntuu9PwH8Mr3/qPuSvLa1eQswfnbyBeA/AA+2s4a99C6J3cbMbAZ+M8nhAEl+KsmR9I6vX233Sobo3SQftws4pU3/6iTLfR5P/Z651X31bwM/0ff+c8Bvjb9JsrJN3krv+CfJWcDRM9kozdg3gBe2e1PPpH2eNP8OoH2mPFFVT0zQf7L9+EXgTa12Bk/tx4Otb9x4aPxz+4zrf5Kr/zh6ADh2/HMzyeFJXjLIZ+HTKUgeAC5Kcj+9f/Sr++a9G1ia5I+qaiu93+N1D/BZetfPx3fw24APJ7kbb1IuClX1JeBTwFfo7c+tbdZq4I+T3EPv/sn7W/td9Pbt+Gn7bcDjVTXTSwcfA+4DvtRufv4FvTP8T9P7Mwj30bvE+iWeOr7+APjTJNuAH0yy3PcBf53eQwH9vyr8b4E3tJukr6V36XY4vRv999G7GT++jp9NsoPeJa7/PcPt0gxU1f+ld2zdCWwBvto3+3tJvkzv/taFkyziYPvxjHZs/RrwKPDtKdY3PqbHgY8C99L7hmdr3+xrgD9vn3FL6IXMB9tN9bt56snAGX0W/lj+ipQkz6mqsfSe7LoVWNM+kKTO+o6vF9D7D//qqnr0UI9L8ye9J/J+p6q2Ddj/mcAP2u8hfBVwdbtBviA9ne6RzMTaJCvonQKuN0Q0yz7TrjMfAVxmiGgALwJuSPIM4PvA2w/xeA7qx/KMRJI0e55O90gkSYeAQSJJ6sQgkSR1YpBIkjoxSCRJnfx/FEe99rG+xdQAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "lung_expr.de.hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4. Mapping gene identifiers\n",
    "\n",
    "We'll map our Ensembl identifiers to Entrez, as this is the admitted input for the `gseapy` enrichment functions we'll be using later on.\n",
    "There are many tools one can use to map identifiers between databases. Note that these conversions may not always cover the whole dataset due to identifier deprecation or one-to-many mappings.\n",
    "\n",
    "The example below uses [`BioMart`](http://www.ensembl.org/biomart):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23136  ids mapped\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Value</th>\n",
       "      <th>de</th>\n",
       "      <th>entrezID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3656</th>\n",
       "      <td>ENSG00000109436</td>\n",
       "      <td>TBC1D9</td>\n",
       "      <td>-0.754440</td>\n",
       "      <td>0.032011</td>\n",
       "      <td>0.204393</td>\n",
       "      <td>bkgd</td>\n",
       "      <td>23158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1797</th>\n",
       "      <td>ENSG00000088930</td>\n",
       "      <td>XRN2</td>\n",
       "      <td>-0.116049</td>\n",
       "      <td>0.714354</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>bkgd</td>\n",
       "      <td>22803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4101</th>\n",
       "      <td>ENSG00000112812</td>\n",
       "      <td>PRSS16</td>\n",
       "      <td>-0.013214</td>\n",
       "      <td>0.979128</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>bkgd</td>\n",
       "      <td>10279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15920</th>\n",
       "      <td>ENSG00000185862</td>\n",
       "      <td>EVI2B</td>\n",
       "      <td>-1.733936</td>\n",
       "      <td>0.000473</td>\n",
       "      <td>0.036285</td>\n",
       "      <td>downregulated</td>\n",
       "      <td>2124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34116</th>\n",
       "      <td>ENSG00000264204</td>\n",
       "      <td>AGAP7P</td>\n",
       "      <td>1.888521</td>\n",
       "      <td>0.085894</td>\n",
       "      <td>0.304664</td>\n",
       "      <td>bkgd</td>\n",
       "      <td>653268</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneID GeneName    log2FC   P.Value  adj.P.Value  \\\n",
       "3656   ENSG00000109436   TBC1D9 -0.754440  0.032011     0.204393   \n",
       "1797   ENSG00000088930     XRN2 -0.116049  0.714354     1.000000   \n",
       "4101   ENSG00000112812   PRSS16 -0.013214  0.979128     1.000000   \n",
       "15920  ENSG00000185862    EVI2B -1.733936  0.000473     0.036285   \n",
       "34116  ENSG00000264204   AGAP7P  1.888521  0.085894     0.304664   \n",
       "\n",
       "                  de entrezID  \n",
       "3656            bkgd    23158  \n",
       "1797            bkgd    22803  \n",
       "4101            bkgd    10279  \n",
       "15920  downregulated     2124  \n",
       "34116           bkgd   653268  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def add_entrez_column_biomart(df, max_chunk_size=400):\n",
    "    ensembl_ids = list(set(df['GeneID'].tolist()))\n",
    "    entrez_dict = {}\n",
    "\n",
    "    # Connect to the BioMart server\n",
    "    server = BiomartServer(\"http://www.ensembl.org/biomart\")\n",
    "    dataset = server.datasets[\"hsapiens_gene_ensembl\"]\n",
    "\n",
    "    # Split ensembl_ids into chunks if necessary and fetch results\n",
    "    for i in range(0, len(ensembl_ids), max_chunk_size):\n",
    "        chunk = ensembl_ids[i:i + max_chunk_size]\n",
    "        response = dataset.search(\n",
    "            {\n",
    "                \"filters\": {\"ensembl_gene_id\": chunk},\n",
    "                \"attributes\": [\"ensembl_gene_id\", \"entrezgene_id\"],\n",
    "            }\n",
    "        )\n",
    "        if response.status_code == 200:\n",
    "            print(f\"{i // max_chunk_size + 1} chunks sent\", end=\"\\r\", flush=True)\n",
    "            # Parse results\n",
    "            for line in response.iter_lines(decode_unicode=True):\n",
    "                ensembl, entrez = line.split(\"\\t\")\n",
    "                entrez_dict[ensembl] = entrez if entrez else None\n",
    "        else:\n",
    "            print(\n",
    "                response.status_code,\n",
    "                \" - request unsuccessful for chunk starting at index \",\n",
    "                i\n",
    "            )\n",
    "\n",
    "    # Map the entrez IDs to the dataframe\n",
    "    df = df.copy()\n",
    "    df['entrezID'] = df['GeneID'].map(entrez_dict)\n",
    "\n",
    "    return df\n",
    "\n",
    "biomart_df = add_entrez_column_biomart(lung_expr)\n",
    "print(len(set(biomart_df.entrezID)), \" ids mapped\")\n",
    "biomart_df.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The example below uses [`BridgeDB`](https://webservice.bridgedb.org):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_entrez_column_bridgedb(df, chunk_size = 300):\n",
    "    ids = set(df[\"GeneID\"].tolist())\n",
    "    entrez_ids = {}\n",
    "\n",
    "    for i in range(0, len(ids), chunk_size):\n",
    "        chunk = list(ids)[i : i + chunk_size]\n",
    "        response = requests.post(\n",
    "            \"https://webservice.bridgedb.org/Human/xrefsBatch/En\",\n",
    "            headers={\"Content-Type\": \"text/plain\", \"accept\": \"application/json\"},\n",
    "            data=\"\\n\".join(chunk),\n",
    "        )\n",
    "        if response.status_code == 200:\n",
    "            print(f\"{i // chunk_size + 1} chunks sent\", end=\"\\r\", flush=True)\n",
    "            res = response.json()\n",
    "            chunk_entrez_ids = {\n",
    "                ensembl: item.split(\"L:\")[1]\n",
    "                for ensembl in res\n",
    "                for item in res[ensembl][\"result set\"]\n",
    "                if \"L:\" in item\n",
    "            }\n",
    "\n",
    "            entrez_ids.update(chunk_entrez_ids)\n",
    "        else:\n",
    "            print(\n",
    "                response.status_code,\n",
    "                \" - request unsuccessful for chunk starting at index \",\n",
    "                i,\n",
    "                end=\"\\r\",\n",
    "            )\n",
    "\n",
    "    df[\"entrezID\"] = df[\"GeneID\"].map(entrez_ids)\n",
    "    return df\n",
    "\n",
    "bdb_df = add_entrez_column_bridgedb(lung_expr)\n",
    "print(len(set(bdb_df.entrezID)), \" ids mapped\", flush=True)\n",
    "bdb_df.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Which mappings database gave us better coverage?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "biomart_len = len(set(biomart_df.entrezID))\n",
    "bdb_len = len(set(bdb_df.entrezID))\n",
    "print(f\"Biomart: {biomart_len}\\nBridgeDb: {bdb_len}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "if biomart_len > bdb_len:\n",
    "    lung_expr = biomart_df\n",
    "else:\n",
    "    lung_expr = bdb_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Conversions weren't complete. What's our coverage?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> **Notice the unmapped genes will be left out of the analyses!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\n",
    "    round((len(set(lung_expr.GeneID)) - len(set(lung_expr.entrezID))) / \n",
    "          len(set(lung_expr.GeneID)) * 100),\n",
    "    \"%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Enrichment\n",
    "### 3.1. Gene Ontology Enrichment\n",
    "\n",
    "Let's start with Gene Ontology. We'll go with `GO_Biological_Process_2023` for an up-to-date enrichment in biological processes the gene sets are involved in.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For **downregulated** genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dr = gp.enrichr(\n",
    "    gene_list=lung_expr[lung_expr.de == \"downregulated\"].GeneName,\n",
    "    background=lung_expr.GeneName,\n",
    "    gene_sets=\"GO_Biological_Process_2023\",\n",
    "    organism=\"human\",\n",
    "    outdir=\"enrichr_results\",\n",
    "    no_plot=True,\n",
    "    verbose=True,\n",
    "    cutoff=P_CUTOFF,\n",
    ").res2d\n",
    "# Add gene count\n",
    "dr[\"gene_count\"] = [len(row.Genes.split(\";\")) for _, row in dr.iterrows()]\n",
    "# Arbitrary adj.P-Val cutoff\n",
    "dr = dr[dr[\"Adjusted P-value\"] < P_CUTOFF]\n",
    "# View results\n",
    "dr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For **upregulated** genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ur = gp.enrichr(\n",
    "    gene_list=lung_expr[lung_expr.de == \"upregulated\"].GeneName,\n",
    "    background=lung_expr.GeneName,\n",
    "    gene_sets=\"GO_Biological_Process_2023\",\n",
    "    organism=\"human\",\n",
    "    outdir=\"enrichr_results\",\n",
    "    no_plot=True,\n",
    "    verbose=True,\n",
    "    cutoff=0.05,\n",
    ").res2d\n",
    "# Add gene count\n",
    "ur[\"gene_count\"] = [len(row.Genes.split(\";\")) for _, row in ur.iterrows()]\n",
    "# Arbitrary adj.P-Val cutoff\n",
    "ur = ur[ur['Adjusted P-value'] < P_CUTOFF]\n",
    "# View results\n",
    "ur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ur.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dr.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Visualizing GO enrichment results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_enrichment(df, number, type, geneset=\"GO\"):\n",
    "    top = (\n",
    "        df\n",
    "            .sort_values(by=\"Adjusted P-value\", ascending=True)\n",
    "            .head(number)\n",
    "            .sort_values(by=\"gene_count\", ascending=False)\n",
    "    )\n",
    "\n",
    "    # Plot the top 100 terms\n",
    "    plt.figure(figsize=(10, 20))\n",
    "    plt.barh(top['Term'], top['gene_count'], color='skyblue')\n",
    "    plt.xlabel('Gene Count')\n",
    "    plt.ylabel('GO Terms')\n",
    "    plt.title(f\"Top {number} {geneset} Terms by Gene Count, type = {type}, Adj. P-Value < {P_CUTOFF}\")\n",
    "    plt.gca().invert_yaxis()\n",
    "    plt.show()\n",
    "    return plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_enrichment(ur, 100, \"upregulated\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_enrichment(dr, 100, \"downregulated\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overlap?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlaps = [\n",
    "    [\n",
    "        rowi.Term,\n",
    "        len(list(set(rowi.Genes.split(\";\") + rowj.Genes.split(\";\")))),\n",
    "        #list(set(rowi.Genes.split(\";\") + rowj.Genes.split(\";\"))),\n",
    "        len(list(set(rowi.Genes.split(\";\")))),\n",
    "        len(list(set(rowj.Genes.split(\";\") ))),\n",
    "        \n",
    "    ]\n",
    "    for _, rowi in ur.iterrows()\n",
    "    for _, rowj in dr.iterrows()\n",
    "    if rowi.Term == rowj.Term\n",
    "]\n",
    "ol_df = pd.DataFrame(\n",
    "    overlaps, columns=[\"term\", \"genecount\", \"genes_up\", \"genes_down\"]\n",
    ").sort_values(by=\"genecount\", ascending=False)\n",
    "ol_df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total = len(set(dr.Term)) + len(set(ur.Term))\n",
    "print(\n",
    "    f\"{len(ol_df)}/{total} OVERLAPPING GO BIOLOGICAL PROCESS TERMS BETWEEN UPREGULATED AND DOWNREGULATED SETS ({round(len(ol_df)/total*100, 2)}%)\"\n",
    ")\n",
    "ol_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Repeat analysis without overlapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dr_clean = dr[dr[\"Term\"].isin(ol_df.term.to_list())]\n",
    "ur_clean = ur[ur[\"Term\"].isin(ol_df.term.to_list())]\n",
    "dr_clean.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ur_clean.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_enrichment(ur[ur.Term.isin(list(ol_df.term))], 100, \"upregulated, overlapping\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_enrichment(dr[dr.Term.isin(list(ol_df.term))], 100, \"downregulated, overlapping\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 WikiPathways Enrichment \n",
    "That's all well and fine, but we are here for *Pathway Analysis!*  Everything up to this point is basic GO analysis. Building on top of this foundation, let's see what WikiPathways can add. The `gseapy` package includes built-in support for WikiPathways gene sets (see https://maayanlab.cloud/Enrichr/#libraries).\n",
    "\n",
    "Thus, for downregulated genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ewp_dr = gp.enrichr(\n",
    "    gene_list=lung_expr[lung_expr.de == \"downregulated\"].GeneName,\n",
    "    background=lung_expr.GeneName,\n",
    "    gene_sets=\"WikiPathways_2024_Human\",\n",
    "    organism=\"human\",\n",
    "    outdir=\"enrichr_results\",\n",
    "    no_plot=True,\n",
    "    verbose=True,\n",
    "    cutoff=0.05,).res2d\n",
    "# View the results\n",
    "ewp_dr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ewp_dr[\"gene_count\"] = [len(row.Genes.split(\";\")) for _, row in ewp_dr.iterrows()]\n",
    "# Arbitrary adj.P-Val cutoff\n",
    "ewp_dr = ewp_dr[ewp_dr[\"Adjusted P-value\"] < P_CUTOFF]\n",
    "# View results\n",
    "ewp_dr.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And upregulated:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ewp_ur = gp.enrichr(\n",
    "    gene_list=lung_expr[lung_expr.de == \"upregulated\"].GeneName,\n",
    "    background=lung_expr.GeneName,\n",
    "    gene_sets=\"WikiPathways_2024_Human\",\n",
    "    organism=\"human\",\n",
    "    outdir=\"enrichr_results\",\n",
    "    no_plot=True,\n",
    "    verbose=True,\n",
    "    cutoff=0.05,\n",
    ").res2d\n",
    "# View the results\n",
    "ewp_ur.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ewp_ur[\"gene_count\"] = [len(row.Genes.split(\";\")) for _, row in ewp_ur.iterrows()]\n",
    "# Arbitrary adj.P-Val cutoff\n",
    "ewp_ur = ewp_dr[ewp_dr[\"Adjusted P-value\"] < P_CUTOFF]\n",
    "# View results\n",
    "ewp_ur.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_enrichment(ewp_dr, 100, \"downregulated\", \"WikiPathways\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_enrichment(ewp_ur, 100, \"upregulated\", \"WikiPathways\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plot_enrichment(ewp_ur, 100, \"upregulated\", \"WikiPathways\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can retrieve the WP ids from the `Term` column:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_wpid_col(df):\n",
    "    wpid = ['WP'+row.Term.split(\"WP\")[1] for _, row in df.iterrows()]\n",
    "    df['wpid'] = wpid\n",
    "    return df\n",
    "\n",
    "ewp_ur, ewp_dr = add_wpid_col(ewp_ur), add_wpid_col(ewp_ur)\n",
    "ewp_ur.sample(4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Explore\n",
    "Turning again to the **pyWikiPathways** package, let's explore the content and some of our pathway hits so far. Since we are studying Lung Cancer here, let's start with a search for relevant pathways..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lc_pathways = pwpw.find_pathways_by_text(\"lung cancer\")\n",
    "lc_pathways = lc_pathways[lc_pathways.species == \"Homo sapiens\"]\n",
    "lc_pathways\n",
    "# Differences with R workflows: find_pathway_names_by_text method DOES NOT ALLOW to filter by species like the R version findPathwayNamesByText, need to use find_pathways_by_text instead"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, so there are just a few human pathways that explicitly mention \"lung cancer\" in their titles or descriptions. How many of these were in our top hits for enrichment?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_ids = pd.concat([ewp_dr, ewp_ur]).wpid\n",
    "common = [\n",
    "    enrichment_id\n",
    "    for i, enrichment_id in enumerate(enrichment_ids)\n",
    "    if enrichment_id in lc_pathways.id.values\n",
    "]\n",
    "print(len(indices), \" pathways:\\n-\", \"\\n- https://wikipathways.org/pathways/\".join(common))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exploring pathways and importing them into Cytoscape with RCy3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = pwpw.get_pathway_info(\"WP179\")\n",
    "print(url['url'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd_list = [\"wikipathways\", \"import-as-pathway\", \"id=WP179\"]\n",
    "cmd = \" \".join(cmd_list)\n",
    "p4c.commands.commands_get(cmd)\n",
    "p4c.notebook_export_show_image()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping gene expression data onto pathways in Cytoscape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p4c.load_table_data(lung_expr, data_key_column=\"GeneID\", table_key_column=\"Ensembl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setting visual styles in Cytoscape\n",
    "Now we can define visual styles to visualize our data on this pathway. First, let's set the node fill color to display the log2 fold change data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p4c.set_node_color_mapping(\"log2FC\", [-1, 0, 1], [\"#67001f\", \"#ffffff\",\"#053061\"], style_name=\"WikiPathways\")\n",
    "# TODO report difference (R--> \"WikiPathways\"; Python: \"WikiPathways-As-Network\"\")\n",
    "p4c.notebook_export_show_image()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can similarly map P-values to border color, etc. There are dozens and dozens of visual properties on nodes and edges available for data visualization!\n",
    "\n",
    "The power of scripting is in doing something multiple times though... So, let's now apply this same data and visual style to the first 5 pathways we are interested in. In just two lines of code...\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for pathway_id in lc_pathways.id[:5]:\n",
    "    cmd = f\"wikipathways import-as-pathway id={pathway_id}\"\n",
    "    p4c.commands.commands_get(cmd)\n",
    "    p4c.load_table_data(lung_expr.copy(), data_key_column=\"GeneID\", table_key_column=\"Ensembl\")\n",
    "    p4c.toggle_graphics_details()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extend\n",
    "Now that our pathways are loaded into Cytoscape, this opens up a ton of potential analysis and visualizations options! Check out the [Cytoscape manual](http://manual.cytoscape.org) and [App Store](http://apps.cytoscape.org/), for starters. You might also browse the [Cytoscape tutorials](http://tutorials.cytoscape.org/) and [RCy3 vignettes](https://bioconductor.org/packages/release/bioc/vignettes/RCy3/inst/doc/Overview-of-RCy3.html) (for R) if you want hands-on examples.\n",
    "\n",
    "In this notebook, we will use the *CyTargetLinker* app for Cytoscape to extend a network representation of our pathway with drug-target interactions.\n",
    "\n",
    "First, let's reimport our pathway *as a network* using this slightly modified command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p4c.commands.commands_get('wikipathways import-as-network id=WP179')\n",
    "p4c.load_table_data(lung_expr, data_key_column=\"GeneID\", table_key_column=\"Ensembl\")\n",
    "p4c.set_node_color_mapping(\"log2FC\", [-1, 0, 1], [\"#67001f\", \"#ffffff\", \"#053061\"], style_name=\"WikiPathways-As-Network\")\n",
    "p4c.notebook_export_show_image()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See the difference?  Same data, same pathway source, but different representation. The network view of pathways is useful when you want to add more nodes, traverse paths, perform automatic layouts, etc.\n",
    "\n",
    "Next, we need to load the latest drug-target database. The databases supported by **CyTargetLinker** are called *linksets* and can be downloaded from the [CyTargetLinker website](https://projects.bigcat.unimaas.nl/cytargetlinker/linksets/). We have provided an example **drugbank** linkset for this vignette, so you don't have to download anything."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unzip(\n",
    "    system.file(\"extdata\", \"drugbank-5.1.0.xgmml.zip\", package=\"rWikiPathways\"),\n",
    "    exdir=getwd(),\n",
    ")\n",
    "drugbank < -file.path(getwd(), \"drugbank-5.1.0.xgmml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have our *drugbank* linkset  loaded, we can run CyTargetLinker as a command:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "commandsRun(\n",
    "    paste0('cytargetlinker extend idAttribute=\"Ensembl\" linkSetFiles=\"', drugbank, '\"')\n",
    ")\n",
    "commandsRun('cytargetlinker applyLayout network=\"current\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This returns information about what was added. And in Cytoscape, you now have a copy of your original network, but now with additional nodes and edges. Check it out...\n",
    "\n",
    "...Hmm, they are kind of plain and hard to see. Let's use Cytoscape visualization styles to fix that!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my.drugs <- selectNodes(\"drug\", by.col = \"CTL.Type\", preserve = FALSE)$nodes #easy way to collect node SUIDs by column value\n",
    "clearSelection()\n",
    "setNodeColorBypass(my.drugs, \"#DD99FF\")\n",
    "setNodeShapeBypass(my.drugs, \"hexagon\")\n",
    "\n",
    "drug.labels <- getTableColumns(columns=c(\"SUID\",\"CTL.label\"))\n",
    "drug.labels <- na.omit(drug.labels)\n",
    "mapply(function(x,y) setNodeLabelBypass(x,y), drug.labels$SUID, drug.labels$CTL.label)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
